libslide/src/parser/expression_pattern_parser.rs

```diff
@@ -101,4 +119,30 @@ mod tests {
                                                     r#"The variable "b" cannot be used in an expression pattern. "#,
                                                     r##"Consider using "$b", "#b", or "_b" as a variable instead."##)
     }
+
+    #[test]
+    fn common_subexpression_elimination() {
+        let program = "$v * #c + $v * #c";
+        let tokens = scan(program);
+        let (parsed, _) = parse(tokens);
+        let (l, r) = match parsed {
+            ExprPat::BinaryExpr(BinaryExpr { lhs, rhs, .. }) => (lhs, rhs),
+            _ => unreachable!(),
+        };
+        assert!(std::ptr::eq(l.as_ref(), r.as_ref())); // $v * #c
+
+        let (ll, lr, rl, rr) = match (l.as_ref(), r.as_ref()) {
+            (
+                ExprPat::BinaryExpr(BinaryExpr {
+                    lhs: ll, rhs: lr, ..
+                }),
+                ExprPat::BinaryExpr(BinaryExpr {
+                    lhs: rl, rhs: rr, ..
+                }),
+            ) => (ll, lr, rl, rr),
+            _ => unreachable!(),
+        };
+        assert!(std::ptr::eq(ll.as_ref(), rl.as_ref())); // 1
+        assert!(std::ptr::eq(lr.as_ref(), rr.as_ref())); // 2
+    }
 }
```

Comment:
This may be just an idea, but can't we declare a few macros here for finish_expr to avoid writing the same function twice in expression_pattern_parser, and expression_parser?

Reply:
I would actually prefer not doing that because the macro would have to live in a file different from both of these, which I don't think is worth doing for just one macro and such a small function. Let me know what you think.

Reply:
thats fine, these will probably be edited anyway